src/core/trtserver.cc

```diff
@@ -311,6 +487,63 @@ TRTSERVER_ServerModelStatus(
   return TrtServerError::Create(request_status.code(), request_status.msg());
 }
 
+TRTSERVER_Error*
+TRTSERVER_ServerInferAsync(
+    TRTSERVER_Server* server,
+    TRTSERVER_InferenceRequestProvider* request_provider,
+    TRTSERVER_InferenceCompleteFn_t complete_fn, void* userp)
+{
+  ni::InferenceServer* lserver = reinterpret_cast<ni::InferenceServer*>(server);
+  TrtServerRequestProvider* lprovider =
+      reinterpret_cast<TrtServerRequestProvider*>(request_provider);
+  ni::InferRequestHeader* request_header = lprovider->InferRequestHeader();
+
+  auto infer_stats = std::make_shared<ni::ModelInferStats>(
+      lserver->StatusManager(), lprovider->ModelName());
+  auto timer = std::make_shared<ni::ModelInferStats::ScopedTimer>();
+  infer_stats->StartRequestTimer(timer.get());
+  infer_stats->SetRequestedVersion(lprovider->ModelVersion());
+  infer_stats->SetFailed(true);
+
+  std::shared_ptr<ni::InferenceServer::InferBackendHandle> backend = nullptr;
+  RETURN_IF_STATUS_ERROR(ni::InferenceServer::InferBackendHandle::Create(
+      lserver, lprovider->ModelName(), lprovider->ModelVersion(), &backend));
+  infer_stats->SetMetricReporter(
+      backend->GetInferenceBackend()->MetricReporter());
+  infer_stats->SetBatchSize(request_header->batch_size());
+
+  RETURN_IF_STATUS_ERROR(ni::NormalizeRequestHeader(
+      *backend->GetInferenceBackend(), *request_header));
+
+  std::shared_ptr<ni::InferRequestProvider> infer_request_provider;
+  RETURN_IF_STATUS_ERROR(ni::InferRequestProvider::Create(
+      lprovider->ModelName(), lprovider->ModelVersion(), *request_header,
+      lprovider->InputMap(), &infer_request_provider));
+
+  std::shared_ptr<ni::DelegatingInferResponseProvider> infer_response_provider;
+  RETURN_IF_STATUS_ERROR(ni::DelegatingInferResponseProvider::Create(
+      *request_header, backend->GetInferenceBackend()->GetLabelProvider(),
+      &infer_response_provider));
+
+  auto request_status = std::make_shared<ni::RequestStatus>();
+  lserver->HandleInfer(
+      request_status.get(), backend, infer_request_provider,
+      infer_response_provider, infer_stats,
+      [infer_stats, timer, infer_response_provider, server, request_status,
+       complete_fn, userp]() mutable {
+        infer_stats->SetFailed(false);
```

Comment:
`infer_stats` will be set in [HandleInfer()](https://github.com/NVIDIA/tensorrt-inference-server/blob/81bbd906e73f6bf3604604d8bcc8e9eefc2af214/src/core/server.cc#L350). And the response may need to include request status so that user can check if the inference is successful.

Reply:
Added TRTSERVER_InferenceResponse* methods to get status and response header.